flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/physical/stream/WatermarkAssignerChangelogNormalizeTransposeRule.java

```diff
@@ -67,18 +84,96 @@ public void onMatch(RelOptRuleCall call) {
             final StreamPhysicalCalc calc = call.rel(1);
             final StreamPhysicalChangelogNormalize changelogNormalize = call.rel(2);
             final StreamPhysicalExchange exchange = call.rel(3);
-
-            final RelNode newTree =
-                    buildTreeInOrder(
-                            changelogNormalize, exchange, watermark, calc, exchange.getInput());
-            call.transformTo(newTree);
+            final Mappings.TargetMapping calcMapping = buildCalcMapping(calc);
```

Comment:
the added code can be extracted into a new method. then the `onMatch` method will be more clean.

Reply:
good point